File: ios/Sources/ACRendering/Views/FactSetView.swift
// Copyright (c) Microsoft Corporation. All rights reserved.
// Author: Vikrant Singh (github.com/VikrantSingh01)
// Licensed under the MIT License.

import SwiftUI
import ACCore
import ACAccessibility
import ACMarkdown

struct FactSetView: View {
    let factSet: FactSet
    let hostConfig: HostConfig

    var body: some View {
        VStack(alignment: .leading, spacing: CGFloat(hostConfig.factSet.spacing)) {
            ForEach(factSet.facts) { fact in
                HStack(alignment: .top, spacing: 8) {
                    factText(fact.title, config: hostConfig.factSet.title)
                        .lineLimit(hostConfig.factSet.title.wrap ? nil : 1)
                        .frame(width: titleMaxWidth > 0 ? CGFloat(titleMaxWidth) : nil, alignment: .leading)
                    factText(fact.value, config: hostConfig.factSet.value)
                        .lineLimit(hostConfig.factSet.value.wrap ? nil : 1)
                        .frame(maxWidth: .infinity, alignment: .leading)
                }
            }
        }
        .spacing(factSet.spacing, hostConfig: hostConfig)
        .separator(factSet.separator, hostConfig: hostConfig)
        .accessibilityContainer(label: "Fact Set")
    }

    /// Renders fact text with markdown support (bold, italic, links) when markdown is detected.
    @ViewBuilder
    private func factText(_ text: String, config: FactSetTextConfig) -> some View {
        let expanded = DateTimeMacroExpander.expand(text)
        if expanded.containsMarkdown {
            let tokens = MarkdownParser.parse(expanded)
            let attributed = MarkdownRenderer.render(
                tokens: tokens,
                font: resolveFont(config),
                color: resolveColor(config)
            )
            Text(attributed)
        } else {
            Text(expanded)
                .font(resolveFont(config))
                .fontWeight(resolveWeight(config.weight))
                .foregroundColor(resolveColor(config))
        }
    }

    private var titleMaxWidth: Int {
        hostConfig.factSet.title.maxWidth
    }

    private func resolveWeight(_ weightString: String) -> Font.Weight {
        let weightValue: Int
        switch weightString.lowercased() {
        case "lighter":
            weightValue = hostConfig.fontWeights.lighter
        case "bolder":
            weightValue = hostConfig.fontWeights.bolder
        default:
            weightValue = hostConfig.fontWeights.default
        }

        switch weightValue {
        case 100...199: return .ultraLight
        case 200...299: return .light
        case 300...399: return .regular
        case 400...499: return .regular
        case 500...599: return .medium
        case 600...699: return .semibold
        case 700...799: return .bold
        default: return .heavy
        }
    }

    private func resolveFont(_ config: FactSetTextConfig) -> Font {
        let size = CGFloat(resolveFontSize(config.size))
        if config.fontType.lowercased() == "monospace" {
            return .system(size: size, design: .monospaced)
        }
        return .system(size: size)
    }

    private func resolveColor(_ config: FactSetTextConfig) -> Color {
        let styleConfig = hostConfig.containerStyles.default
        let colorConfig: ColorConfig

        switch config.color.lowercased() {
        case "dark": colorConfig = styleConfig.foregroundColors.dark
        case "light": colorConfig = styleConfig.foregroundColors.light
        case "accent": colorConfig = styleConfig.foregroundColors.accent
        case "good": colorConfig = styleConfig.foregroundColors.good
        case "warning": colorConfig = styleConfig.foregroundColors.warning
        case "attention": colorConfig = styleConfig.foregroundColors.attention
        default: colorConfig = styleConfig.foregroundColors.default
        }

        let hex = config.isSubtle ? colorConfig.subtle : colorConfig.default
        return Color(hex: hex)
    }

    private func resolveFontSize(_ sizeString: String) -> Int {
        switch sizeString.lowercased() {
        case "small": return hostConfig.fontSizes.small
        case "medium": return hostConfig.fontSizes.medium
        case "large": return hostConfig.fontSizes.large
        case "extralarge": return hostConfig.fontSizes.extraLarge
        default: return hostConfig.fontSizes.`default`
        }
    }
}

File: android/ac-rendering/src/main/kotlin/com/microsoft/adaptivecards/rendering/composables/FactSetView.kt
// Copyright (c) Microsoft Corporation. All rights reserved.
// Author: Vikrant Singh (github.com/VikrantSingh01)
// Licensed under the MIT License.

package com.microsoft.adaptivecards.rendering.composables

import androidx.compose.foundation.layout.*
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.microsoft.adaptivecards.core.models.FactSet
import com.microsoft.adaptivecards.rendering.theme.LocalHostConfig
import com.microsoft.adaptivecards.accessibility.scaledTextSize
import com.microsoft.adaptivecards.markdown.MarkdownParser
import com.microsoft.adaptivecards.markdown.MarkdownRenderer
import com.microsoft.adaptivecards.markdown.containsMarkdown
// DateTimeMacroExpander is in the same package

/**
 * Renders a FactSet element as key-value pairs.
 * Resolves font size, weight, and color from HostConfig factSet configuration.
 */
@Composable
fun FactSetView(
    element: FactSet,
    modifier: Modifier = Modifier
) {
    val hostConfig = LocalHostConfig.current
    val titleSize = resolveFontSize(hostConfig.factSet.title.size, hostConfig)
    val valueSize = resolveFontSize(hostConfig.factSet.value.size, hostConfig)
    val titleLineHeight = resolveLineHeight(hostConfig.factSet.title.size, hostConfig).sp
    val valueLineHeight = resolveLineHeight(hostConfig.factSet.value.size, hostConfig).sp
    val titleWeight = resolveFontWeight(hostConfig.factSet.title.weight, hostConfig)
    val valueWeight = resolveFontWeight(hostConfig.factSet.value.weight, hostConfig)
    val titleColor = getTextColor(hostConfig.factSet.title.color, hostConfig.factSet.title.isSubtle, hostConfig)
    val valueColor = getTextColor(hostConfig.factSet.value.color, hostConfig.factSet.value.isSubtle, hostConfig)
    val titleMaxWidth = hostConfig.factSet.title.maxWidth

    Column(
        modifier = modifier,
        verticalArrangement = Arrangement.spacedBy(hostConfig.factSet.spacing.dp)
    ) {
        element.facts.forEach { fact ->
            Row(
                horizontalArrangement = Arrangement.spacedBy(8.dp)
            ) {
                // Title (key) — use fixed max width, wrap content
                val titleText = DateTimeMacroExpander.expand(fact.title)
                if (titleText.containsMarkdown()) {
                    val tokens = MarkdownParser.parse(titleText)
                    val annotated = MarkdownRenderer.render(tokens, titleSize.sp, titleColor)
                    Text(
                        text = annotated,
                        fontWeight = titleWeight,
                        fontSize = scaledTextSize(titleSize),
                        lineHeight = titleLineHeight,
                        fontFamily = resolveFontFamily(hostConfig.factSet.title.fontType),
                        maxLines = if (hostConfig.factSet.title.wrap) Int.MAX_VALUE else 1,
                        overflow = TextOverflow.Ellipsis,
                        modifier = Modifier.widthIn(
                            max = if (titleMaxWidth > 0) titleMaxWidth.dp else 150.dp
                        )
                    )
                } else {
                    Text(
                        text = titleText,
                        fontWeight = titleWeight,
                        fontSize = scaledTextSize(titleSize),
                        lineHeight = titleLineHeight,
                        color = titleColor,
                        fontFamily = resolveFontFamily(hostConfig.factSet.title.fontType),
                        maxLines = if (hostConfig.factSet.title.wrap) Int.MAX_VALUE else 1,
                        overflow = TextOverflow.Ellipsis,
                        modifier = Modifier.widthIn(
                            max = if (titleMaxWidth > 0) titleMaxWidth.dp else 150.dp
                        )
                    )
                }

                // Value — wrap content, with markdown support
                val valueText = DateTimeMacroExpander.expand(fact.value)
                if (valueText.containsMarkdown()) {
                    val tokens = MarkdownParser.parse(valueText)
                    val annotated = MarkdownRenderer.render(tokens, valueSize.sp, valueColor)
                    Text(
                        text = annotated,
                        fontWeight = valueWeight,
                        fontSize = scaledTextSize(valueSize),
                        lineHeight = valueLineHeight,
                        fontFamily = resolveFontFamily(hostConfig.factSet.value.fontType),
                        maxLines = if (hostConfig.factSet.value.wrap) Int.MAX_VALUE else 1,
                        overflow = TextOverflow.Ellipsis
                    )
                } else {
                    Text(
                        text = valueText,
                        fontWeight = valueWeight,
                        fontSize = scaledTextSize(valueSize),
                        lineHeight = valueLineHeight,
                        color = valueColor,
                        fontFamily = resolveFontFamily(hostConfig.factSet.value.fontType),
                        maxLines = if (hostConfig.factSet.value.wrap) Int.MAX_VALUE else 1,
                        overflow = TextOverflow.Ellipsis
                    )
                }
            }
        }
    }
}

/**
 * Resolve a font type string to a Compose FontFamily
 */
private fun resolveFontFamily(fontType: String): FontFamily {
    return when (fontType.lowercase()) {
        "monospace" -> FontFamily.Monospace
        else -> FontFamily.Default
    }
}
